_posts/2022-10-30-multi-column-sorts-in-arrow-rust-part-2.md

```diff
@@ -0,0 +1,240 @@
+
+However, often values in string columns differ substantially in length or the maximum length is not known at the start of execution, making it inadvisable and/or impractical to pad the strings to a fixed length, and thus the Rust Arrow row format uses a variable length encoding.
+
+We need an encoding that unambiguously terminates the end of the byte array. This not only permits recovering the original value from the row format, but ensures bytes of a longer byte array won’t be compared against bytes from a different field in a row with a shorter byte array during comparison.
+
+A null byte array is encoded as a single `0x00` byte. Similarly, an empty byte array is encoded as a single `0x01` byte.
+
+To encode a non-null, non-empty array, first a single `0x02` byte  is written. Then the array is written in 32-byte blocks, with each block followed by a `0xFF` byte as a continuation token. The final block is padded to 32-bytes with `0x00`, and is then followed by the unpadded length of this final block as a single byte.
+
+Note the following example encodings use a block size of 4 bytes, as opposed to the 32 bytes for brevity
+
+```
+                      ┌───┬───┬───┬───┬───┬───┬───┐
+ "MEEP"               │02 │'M'│'E'│'E'│'P'│FF │04 │
+                      └───┴───┴───┴───┴───┴───┴───┘
+
+                      ┌───┬───┬───┬───┬───┬───┬───┐
+ ""                   │02 │00 │00 │00 │00 │FF │00 │
+                      └───┴───┴───┴───┴───┴───┴───┘
+
+ NULL                 ┌───┬───┬───┬───┬───┬───┬───┐
+                      │00 │00 │00 │00 │00 │FF │00 │
+                      └───┴───┴───┴───┴───┴───┴───┘
+
+"Defenestration"      ┌───┬───┬───┬───┬───┬───┐
+                      │02 │'D'│'e'│'f'│'e'│FF │
+                      └───┼───┼───┼───┼───┼───┤
+                          │'n'│'e'│'s'│'t'│FF │
+                          ├───┼───┼───┼───┼───┤
+                          │'r'│'a'│'t'│'r'│FF │
+                          ├───┼───┼───┼───┼───┤
+                          │'a'│'t'│'i'│'o'│FF │
+                          ├───┼───┼───┼───┼───┼───┐
+                          │'n'│00 │00 │00 │FF │17 │
+                          └───┴───┴───┴───┴───┴───┘
+```
+
+This approach is loosely inspired by [COBS encoding](https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing), and chosen over more traditional [byte stuffing](https://en.wikipedia.org/wiki/High-Level_Data_Link_Control#Asynchronous_framing) as it is more amenable to vectorization, in particular hardware with AVX-256 can copy a 32-byte block in a single instruction.
+
+### Dictionary Arrays
+Dictionary Encoded Data (called [categorical](https://pandas.pydata.org/docs/user_guide/categorical.html) in pandas) is increasingly important because they can store low cardinality data very efficiently in parquet and similar formats.
```

Comment:
```suggestion
Dictionary Encoded Data (called [categorical](https://pandas.pydata.org/docs/user_guide/categorical.html) in pandas) is increasingly important because they can store and process low cardinality data very efficiently.
```